xapian-bindings/python3/Makefile.am

```diff
@@ -99,24 +126,59 @@ xapian/_xapian$(PYTHON3_SO): _xapian.la
 	. ./libtoolconfig.tmp; cp $$objdir/_xapian$(PYTHON3_SO) xapian
 	rm -f libtoolconfig.tmp
 
+# We "import _xapianletor" first so that if we fail to import the glue library
+# we don't generate a broken .pyc or .pyo, and we do it with the current
+# directory set to where the module is as Python with -c always searches the
+# current directory first.
+#
+# We then "import xapianletor" as a separate command, as that's more like how things
+# are once installed - in particular this will fail if the SWIG shadow wrapper
+# doesn't cope with being xapianletor/__init__.py (currently it needs fixing
+# up, which is done by fixup-swig-py3-wrapper).
+xapianletor/__pycache__/__init__.@PYTHON3_CACHE_TAG@.pyc: xapian/__pycache__/__init__.@PYTHON3_CACHE_TAG@.pyc xapianletor/__init__.py xapianletor/_xapianletor$(PYTHON3_SO)
+	$(OSX_SIP_HACK_ENV) $(PYTHON3) -c 'import os;os.chdir("xapianletor");import _xapianletor'
+	$(OSX_SIP_HACK_ENV) $(PYTHON3) -c 'import xapianletor'
+
+xapianletor/__pycache__/__init__.@PYTHON3_CACHE_OPT1_EXT@: xapian/__pycache__/__init__.@PYTHON3_CACHE_OPT1_EXT@ xapianletor/__init__.py xapianletor/_xapianletor$(PYTHON3_SO)
+	$(OSX_SIP_HACK_ENV) $(PYTHON3) -O -c 'import os;os.chdir("xapianletor");import _xapianletor'
+	$(OSX_SIP_HACK_ENV) $(PYTHON3) -O -c 'import xapianletor'
+
+xapianletor/_xapianletor$(PYTHON3_SO): _xapian.la _xapianletor.la
+	$(MKDIR_P) xapianletor
+	$(LIBTOOL) --config > libtoolconfig.tmp
+## ksh requires a path on the sourced file.
+	. ./libtoolconfig.tmp; cp $$objdir/_xapianletor$(PYTHON3_SO) xapianletor
+	rm -f libtoolconfig.tmp
+
 CLEANFILES = \
     xapian/_xapian$(PYTHON3_SO) \
     xapian/__init__.py \
     xapian/__pycache__/__init__.@PYTHON3_CACHE_TAG@.pyc \
     xapian/__pycache__/__init__.@PYTHON3_CACHE_OPT1_EXT@ \
+    xapianletor/_xapianletor$(PYTHON3_SO) \
+    xapianletor/__init__.py \
+    xapianletor/__pycache__/__init__.@PYTHON3_CACHE_TAG@.pyc \
+    xapianletor/__pycache__/__init__.@PYTHON3_CACHE_OPT1_EXT@ \
+    err_output_ListNet_2.txt \
+    ndcg_output_ListNet_2.txt \
+    training_output1.txt \
```

Comment:
This isn't for now, but the fact that we need to do this suggests that the API is slightly too high-level around training and scoring. There should be an API (which is the one we actually test) which returns the data, rather than writing out files. (There are various situations where you don't want to or even can't write to your file system but can store output elsewhere.)

Reply:
I will try to change the API to incorporate the changes you suggested.

Reply:
Don't jump on it right away. Add it to your list of changes (either for during the project or after). It's always good to spot future useful work, but there's an important discipline in not getting distracted by it!